feat(adf-bootstrap): (#472) modify trust relations for roles :zap:

AndreasAugustin opened the pull request:
*Issue #472 *

*Description of changes:*
Modify the trust relations for the roles.
Now a recreation of the related roles is possible without the need of a recreation for the adf stacks in all accounts.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

StewartW commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -34,8 +34,11 @@ Resources:
         Statement:
           - Effect: Allow
+            Condition:
+              StringEquals:
```
```suggestion
              ArnEquals:
```

StewartW commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -201,21 +204,27 @@ Resources:
           - Effect: Allow
             Sid: "AssumeRoleLambda"
+            Condition:
+              StringEquals:
```
```suggestion
              ArnEquals:
```

StewartW commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -201,21 +204,27 @@ Resources:
             Condition:
-              ArnEquals:
-                "aws:SourceArn": !Sub "arn:${AWS::Partition}:codepipeline:${AWS::Region}:${DeploymentAccountId}:*"
+              Fn::And:
```
This doesn't look right here. 
Are you trying to add both conditions in an AND fashion?

StewartW commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -231,9 +240,12 @@ Resources:
           - Effect: Allow
             Sid: "AssumeRole"
+            Condition:
+              StringEquals:
```
```suggestion
              ArnEquals:
```

StewartW commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -333,9 +345,12 @@ Resources:
           - Effect: Allow
             Sid: "AssumeRole"
+            Condition:
+              StringEquals:
```
ArnEquals

StewartW reviewed (changes requested):
I've requested a few changes and a comment. 

Thanks for raising the PR!

StewartW commented:
Hey @AndreasAugustin thanks for raising this PR. It's a great solution. 

Regarding the multi-block condition, is there any reason you wouldn't do it the same way as defined here: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_multi-value-conditions.html 

Rather than using the intrinsic AND operator.

AndreasAugustin commented:
> Hey @AndreasAugustin thanks for raising this PR. It's a great solution.
> 
> Regarding the multi-block condition, is there any reason you wouldn't do it the same way as defined here: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_multi-value-conditions.html
> 
> Rather than using the intrinsic AND operator.

Hi @StewartW thanks a lot for the review.
- fixed: `StringEquals` -> `ArnEquals`
- also moved from single quotes in some statements to double quotes
- `Fn::And` -> migrated to object based. I did not test it. At least `cfn-lint` is not complaining :+1:  I try to test it ASAP

sbkok commented on src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/global.yml:
```diff
@@ -199,16 +202,27 @@ Resources:
                 - cloudformation.amazonaws.com
             Action:
               - sts:AssumeRole
+          - Effect: Allow
+            Sid: "AssumeRoleLambda"
+            Principal:
+              AWS:
+                - !Sub arn:${AWS::Partition}:iam::${DeploymentAccountId}:root
+            Action:
+              - sts:AssumeRole
+            Condition:
+              ArnEquals:
+                "aws:PrincipalArn": !Sub "arn:${AWS::Partition}:iam::${DeploymentAccountId}:role/adf-lambda-role"
```
My mistake while merging, this is no longer in use. It got replaced by the the adf-automation roles instead.
```suggestion
```

sbkok reviewed (dismissed):
Great fix, thanks for contributing this!
